bounded/src/bdd.rs

```diff
@@ -168,21 +167,40 @@ impl Context<'_> {
             valuations[i] = Some(bdd.sat_witness().unwrap());
         }
 
-        println!("found counterexample!");
-        for (i, valuation) in valuations.iter().enumerate() {
-            println!("state {}:", i);
-            let valuation = valuation.clone().unwrap();
-            for relation in self.indices.keys().sorted() {
-                print!("{}: {{", relation);
-                for (elements, (i, _mutable)) in self.indices[relation].iter().sorted() {
-                    if valuation.value(self.vars[*i]) {
-                        print!("{:?}, ", elements);
-                    }
-                }
-                println!("}}");
-            }
-            println!();
-        }
+        // Convert valuations to models
+        let valuations: Vec<_> = valuations.into_iter().map(Option::unwrap).collect();
+        let universe = self
+            .signature
+            .sorts
+            .iter()
+            .map(|s| self.universe[s])
+            .collect();
+        valuations
+            .into_iter()
+            .map(|valuation| {
+                Model::new(
+                    self.signature,
+                    &universe,
+                    self.signature
+                        .relations
+                        .iter()
+                        .map(|r| {
+                            let shape = r
+                                .args
+                                .iter()
+                                .map(|s| cardinality(self.universe, s))
+                                .chain([2])
+                                .collect();
+                            Interpretation::new(&shape, |elements| {
+                                valuation
+                                    .value(self.vars[self.indices[r.name.as_str()][elements].0])
```

Comment:
I think this expression starting with `self.vars[...]` is a good candidate for a method on `self`.

Reply:
Why? It's not used anywhere else, and `Context::get` does what is normally desired. This is the one place in the code where we convert `BddValuation`s to `Model`s.

Reply:
The reason to create a function in this case would mainly be to name and document it (which is helpful even if used only once). Now that you've factored out `convert_valuation` I agree this isn't as necessary.